File: Cpp/Waktu/Time.cpp
#include <iostream>
#include <iomanip>

class Time {
private:
    int second, minute, hour;

public:
    // Constructors
    Time(int second, int minute, int hour) : second(second), minute(minute), hour(hour) {}
    Time() : second(0), minute(0), hour(0) {}

    // Getters
    int getSecond() const { return second; }
    int getMinute() const { return minute; }
    int getHour() const { return hour; }

    // Setters
    void setSecond(int second) { this->second = second; }
    void setMinute(int minute) { this->minute = minute; }
    void setHour(int hour) { this->hour = hour; }

    // toString method
    std::string toString() const {
        return std::to_string(hour) + ":" + std::to_string(minute) + ":" + std::to_string(second);
    }

    // setTime method
    void setTime(int second, int minute, int hour) {
        this->second = second;
        this->minute = minute;
        this->hour = hour;
    }

    // nextSecond method
    Time& nextSecond() {
        ++second;
        if (second >= 60) {
            second = 0;
            ++minute;
            if (minute >= 60) {
                minute = 0;
                ++hour;
                if (hour >= 24) {
                    hour = 0;
                }
            }
        }
        return *this; // Return "*this" instance, to support chaining operations
    }
};


int main() {
    // Test Constructors and toString()
    Time t1(1, 2, 3);
    std::cout << t1.toString() << std::endl;

    Time t2; // Default constructor
    std::cout << t2.toString() << std::endl;

    // Test Setters and Getters
    t1.setHour(4);
    t1.setMinute(5);
    t1.setSecond(6);
    std::cout << t1.toString() << std::endl;
    std::cout << "Jamnya adalah: " << t1.getHour() << std::endl;
    std::cout << "Menitnya adalah: " << t1.getMinute() << std::endl;
    std::cout << "Detiknya adalah: " << t1.getSecond() << std::endl;

    // Test setTime()
    t1.setTime(58, 59, 23);
    std::cout << t1.toString() << std::endl;

    // Test nextSecond() and chaining
    std::cout << t1.nextSecond().toString() << std::endl; // Return an instance of Time. Invoke Time's toString()
    std::cout << t1.nextSecond().nextSecond().nextSecond().toString() << std::endl;  // chaining



    return 0;
}

File: JAVA/Time/Time.java
/**
 * Kelas Waktu memodelkan instance waktu dengan detik, menit, dan jam.
 * Kelas ini tidak melakukan validasi input untuk detik, menit, dan jam.
 */
public class Time {
   // The private instance variables
   private int second, minute, hour;

   // The constructors (overloaded)
   /** Membuat instance Waktu dengan detik, menit, dan jam yang diberikan. Tidak ada validasi masukan */
   public Time(int second, int minute, int hour) {
      this.second = second;
      this.minute = minute;
      this.hour = hour;
   }
   /** Membuat instance Time dengan nilai default */
   public Time() {  // the default constructor
      this.second = 0;
      this.minute = 0;
      this.hour = 0;
   }

   // The public getters/setters for the private variables.
   /** Mengambalikan detik */
   public int getSecond() {
      return this.second;
   }
   /** Mengembalikan menit */
   public int getMinute() {
      return this.minute;
   }
   /** Mengembalikan jam */
   public int getHour() {
      return this.hour;
   }
   /** Mengatur detik. No input validation */
   public void setSecond(int second) {
      this.second = second;
   }
   /** Mengatur menit. No input validation */
   public void setMinute(int minute) {
      this.minute = minute;
   }
   /** Mengatur jam. No input validation */
   public void setHour(int hour) {
      this.hour = hour;
   }

   /** Mengembalikan string deskriptif mandiri dalam bentuk "hh: mm: dd" dengan nol di depan */
   public String toString() {
        // Gunakan fungsi bawaan String.format() untuk membentuk String yang diformat
        return String.format("%02d:%02d:%02d", hour, minute, second);
              // Penentu "0" untuk mencetak angka nol di depan, jika tersedia.
   }

   /** Menetapkan detik, menit, dan jam ke nilai tertentu */
   public void setTime(int second, int minute, int hour) {
      // No input validation
      this.second = second;
      this.minute = minute;
      this.hour = hour;
   }

   /** Memajukan instance Waktu ini satu detik, dan mengembalikan instance ini untuk mendukung perantaian*/
   public Time nextSecond() {
      ++second;
      if (second >= 60) {
         second = 0;
         ++minute;
         if (minute >= 60) {
            minute = 0;
            ++hour;
            if (hour >= 24) {
               hour = 0;
            }
         }
      }
      return this;   // Return "this" instance, to support chaining operations
                     // e.g., t1.nextSecond().nextSecond()
   }
}
